### PR TITLE
Research GitHub API rate limits

### PR DESCRIPTION
**Pull request by csomh**
Calculate the approximate number of calls currently made in order to be able to asses the current situation and tell the impact of the proposed approaches. [Direct link to the rendered Jupyter Notebook](https://github.com/csomh/research/blob/github-api-call-inventory/github-api-call-inventory/github-api-call-inventory.ipynb).

Propose approaches to collect metrics.

Related to packit/packit-service#1765

Signed-off-by: Hunor Csomortáni <csomh@redhat.com>

### CODE REVIEW

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [inline github-api-call-inventory/github-api-call-inventory.ipynb:24]: Is the GitHub API called for any other reason while handling PR events?

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [inline github-api-call-inventory/github-api-call-inventory.ipynb:106]: Are there any other events received during a pipeline run?

Comment (lbarcziova) [reply github-api-call-inventory/github-api-call-inventory.ipynb:106]: I think "SRPM build in Copr was submitted..." is the only missing

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (lbarcziova) [reply github-api-call-inventory/github-api-call-inventory.ipynb:24]: iirc we also check whether the one who triggered the event has write access to repo or is author of the PR

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb:106]: Is that a separate event, or that's the status which is set when the update event for the PR is received?

**Review by jpopelka — COMMENTED** [review]
:+1:

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb:24]: Where is this happening? I've looked in the packit-service code, but I was able to find such check only for vm_image jobs. What am I missing? :thinking:

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (lbarcziova) [reply github-api-call-inventory/github-api-call-inventory.ipynb:106]: it is not a separate event, but it happens in a new task:
we receive a Github PR-related event, we create the "Task was accepted status"  ( + do the config loading, private check,..) and then create a new specific task and there we create the "SRPM build in Copr was submitted..." status

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb:24]: Found [it](https://github.com/packit/packit-service/blob/b946857a437b062d6e93021b5420bf46f3ce9964/packit_service/worker/jobs.py#L385).

Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb]: Added.

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb]: Thanks, updated this.

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [inline github-api-call-inventory/github-api-call-inventory.ipynb:458]: Is this observation correct? Would be nice if someone else could confirm this, b/c otherwise the numbers are way too wrong :sweat_smile:

Comment (TomasTomecek) [reply github-api-call-inventory/github-api-call-inventory.ipynb:458]: they shouldn't change within a pipeline run, right? people can become collaborators, be approved, packit.yaml can change, but NOT within a pipeline run: +1 to cache all of that data

this will also be a massive speedup

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (lbarcziova) [inline github-api-call-inventory/github-api-call-inventory.ipynb:101]: maybe I am mistaken, but I think the user check is done only once

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (lbarcziova) [inline github-api-call-inventory/github-api-call-inventory.ipynb:724]: agreed, I would not try to reduce these for the cost of worse UX

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (lbarcziova) [inline github-api-call-inventory/github-api-call-inventory.ipynb:550]: I'm not sure about the UX here, but this could somehow help with the situation with GitLab statuses where we cannot overwrite "running" -> "running" status (packit/packit-service#1821): this way, we could have only "In progress" status pointing to the build/test correctly

**Review by lbarcziova — APPROVED** [review]
very nice write-up, thanks for all the numbers, this can definitely help with prioritising the improvements!

I think it would make sense to start with optimisations for packit.yaml loading and private check

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb:101]: Right, Copr, Koji and TestingFarm event's [don't trigger this check](https://github.com/packit/packit-service/blob/bc1f66bb82aa7c013bde8b8e670ce7e02f55de41/packit_service/worker/allowlist.py#L237). Thanks for pointing it out!

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb:550]: If GitLab has a "pending" status, and supports updates to that, then yes, this could help. Then the "running" status would be set only when the test is actually triggered.

Comment (lbarcziova) [reply github-api-call-inventory/github-api-call-inventory.ipynb:550]: yes, GitLab has a "pending" status

**Comment by csomh** [comment]
I wrote down some ideas on collecting metrics on GitHub API Calls. This is ready for final review and approval now.

File: github-api-call-inventory/README.md
Comment (csomh) [inline github-api-call-inventory/README.md:32]: Would be nice to get a second pair of eyes on this if it indeed makes sense :smiley:

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (lachmanfrantisek) [inline github-api-call-inventory/github-api-call-inventory.ipynb:687]: Agree, this can be a big help, but a) not easy to do b) not always wanted (e.g. with branch commit push, I don't think it's clear if the user wants to skip the build triggered by a previous commit in case of it still being in progress).

Do you think it's worth investigating? Or, it's more beneficial to spend our time on other suggestions you've found?

File: github-api-call-inventory/README.md
Comment (lachmanfrantisek) [inline github-api-call-inventory/README.md:23]: @mfocko Do you think we will be able to implement such feature into OGR?

If we don't need to be precise, I think we can try collecting function calls of pygithub -- no 1-1 mapping to remote calls, but at least a rough idea of where to look closer.

File: github-api-call-inventory/README.md
Comment (lachmanfrantisek) [reply github-api-call-inventory/README.md:32]: Will this "man-in-the-middle" work?

If yes, then it looks like the cleanest solution without any unreliable library patching...

File: github-api-call-inventory/README.md
Comment (lachmanfrantisek) [reply github-api-call-inventory/README.md:23]: It reminds me of our beginnings with ogr and requre when we were patching off all the calls... Actually, we can try using requre for this since there is a way how to catch all the remote calls on `requests` level....;)

**Review by lachmanfrantisek — APPROVED** [review]
Wow! Thanks for such a detailed investigation! I'm really glad you haven't done any hacky solution for the original issue instead of this! Thanks!

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb:469]: I'm not suggesting anything :hand_over_mouth: But yes, probably that'll be the solution.

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb:510]: Currently when build events are handled both build flags (ex: `rpm-build:fedora-37-x86_64`) and test flags (ex: `testing-farm:fedora-37-x86_64`) are updated.

This means that test flags will cycle through the same 

1. "SRPM build in Copr was submitted"
1. "SRPM build in Copr has started..."
1. "SRPM build in Copr has finished."
2. "RPM build is in progress..."
3. "RPMs were built successfully."

messages. Setting each of these is another API call. Instead of doing this when the SRPM build in Copr is submitted, test flags could be set to something like "Waiting for RPMs to be built...", and updated only when tests are submitted.

Comment (lachmanfrantisek) [reply github-api-call-inventory/github-api-call-inventory.ipynb:510]: Yes, that's what I meant. Thanks and agree.

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (csomh) [reply github-api-call-inventory/github-api-call-inventory.ipynb:687]: Why would it be difficult?

We have pipelines and from those we should know which builds/tests are still running. We could cancel those, by implementing something like `pipeline.cancel()`. 

Canceling previous pipelines could be optional, or not done at all in the case of branch pushes. In the case of PRs though it doesn't make sense to let old pipelines to finish.

File: github-api-call-inventory/README.md
Comment (csomh) [reply github-api-call-inventory/README.md:32]: I'll try to come up with a small local experiment to try this out.

File: github-api-call-inventory/README.md
Comment (TomasTomecek) [inline github-api-call-inventory/README.md:13]: I'd also appreciate seeing the status codes of those requests so we know how many of them are `ok`, `4xx`, `2xx`

File: github-api-call-inventory/README.md
Comment (TomasTomecek) [reply github-api-call-inventory/README.md:32]: whaaaat, I would love to see this experiment 😍

if it proves reliable, we could use it for other services as well

**Review by TomasTomecek — APPROVED** [review]

File: github-api-call-inventory/README.md
Comment (csomh) [reply github-api-call-inventory/README.md:32]: It seems the proxy approach would be more complicated than I assumed. NGINX doesn't support forward proxying out of the box (in this case, this is what is needed, not the reverse proxy feature). Solutions do exist, but they involve patching NGINX code, which I consider quite hacky.

I've started looking at PyGithub again, and searching for a way to inject some response hooks into the `requests` session in `github.Requester`, and figure out a way to reliably pass down these hooks from ogr objects. Metrics collections could happen in these hook functions, and we could also make @TomasTomecek happy and create metrics for (groups of) status codes :slightly_smiling_face:

**Comment by csomh** [comment]
I've updated this to describe a rough approach for how to patch PyGithub in order to be able to use a metric collection object for counting calls. Please take another look at "Approach 1".

**Review by TomasTomecek — APPROVED** [review]
Wow, the numbers are just mind-blowing.

File: github-api-call-inventory/github-api-call-inventory.ipynb
Comment (lachmanfrantisek) [reply github-api-call-inventory/github-api-call-inventory.ipynb:687]: (Sorry for not getting back to this earlier..)

> Why would it be difficult?

I am not saying difficult, just that it requires a non-trivial change. (Compared to other enhancements that are more isolated.)

Comment (lachmanfrantisek) [reply github-api-call-inventory/github-api-call-inventory.ipynb:687]: So, not against it, but not sure if we should spend our time implementing that...;)

File: github-api-call-inventory/README.md
Comment (lachmanfrantisek) [inline github-api-call-inventory/README.md:37]: Sounds good!

Comment (lachmanfrantisek) [reply github-api-call-inventory/README.md:37]: What do you suggest from the deployment point of view?

* A PyGithub source-git?
* Do the patching inside image-build?

Maybe, we can try to submit this upstream since this might be relevant for others as well... But I would expect if this got refused...

**Review by lachmanfrantisek — APPROVED** [review]
Thanks again Hunor!

File: github-api-call-inventory/README.md
Comment (csomh) [reply github-api-call-inventory/README.md:37]: I would apply the patches inside the image build, to keep things simple. And yes, they could be submitted upstream, to get community feedback on it.

Comment (csomh) [reply github-api-call-inventory/README.md:37]: Made a note about this.

Comment (lachmanfrantisek) [reply github-api-call-inventory/README.md:37]: Thanks!

**Pull request merged**
